### PR TITLE
8282723: Add constructors taking a cause to JSSE exceptions

### PR DESCRIPTION
**Pull request by XueleiFan**
Please review this small API enhancement to add the usual constructors taking a cause to javax.net.ssl exceptions.  The use of initCause in the JSSE implementation code is updated to use the new constructors accordingly.

Please review the CSR: https://bugs.openjdk.java.net/browse/JDK-8282724

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8282723](https://bugs.openjdk.java.net/browse/JDK-8282723): Add constructors taking a cause to JSSE exceptions
 * [JDK-8282724](https://bugs.openjdk.java.net/browse/JDK-8282724): Add constructors taking a cause to JSSE exceptions (**CSR**)

### Reviewers
 * [Bradford Wetmore](https://openjdk.java.net/census#wetmore) (@bradfordwetmore - **Reviewer**) ⚠️ Review applies to 61cc7934c7e541ffabc42a53c0b2a3f6a24cf869
 * [Iris Clark](https://openjdk.java.net/census#iris) (@irisclark - **Reviewer**) ⚠️ Review applies to 61cc7934c7e541ffabc42a53c0b2a3f6a24cf869

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7722/head:pull/7722` \
`$ git checkout pull/7722`

Update a local copy of the PR: \
`$ git checkout pull/7722` \
`$ git pull https://git.openjdk.java.net/jdk pull/7722/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7722`

View PR using the GUI difftool: \
`$ git pr show -t 7722`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7722.diff">https://git.openjdk.java.net/jdk/pull/7722.diff</a>

</details>

### CODE REVIEW

**Comment by dfuch** [comment]
LGTM. Someone from security-dev should probably review this too.

File: src/java.base/share/classes/javax/net/ssl/SSLException.java
Comment (bradfordwetmore) [inline src/java.base/share/classes/javax/net/ssl/SSLException.java:52]: Thanks for changing the style.  I've always hated the extra vertical space.  :)

**Comment by rhalade** [comment]
Please consider adding simple tests to cover new APIs, similar to https://github.com/openjdk/jdk/pull/7705/files#diff-28e1041be519b6266b3b92aea29c5d8917c279880da7e5e9823a518196e96ea5

**Comment by rhalade** [comment]
Update following for SSLPeerUnverifiedException -

https://github.com/openjdk/jdk/blob/master/src/java.naming/share/classes/com/sun/jndi/ldap/ext/StartTlsResponseImpl.java#L439

File: src/java.base/share/classes/sun/security/ssl/SSLTrafficKeyDerivation.java
Comment (bradfordwetmore) [inline src/java.base/share/classes/sun/security/ssl/SSLTrafficKeyDerivation.java:158]: I can't quite tell given the coloring, but did this get changed into a tab?

File: src/java.base/share/classes/sun/security/ssl/ECDHKeyExchange.java
Comment (bradfordwetmore) [inline src/java.base/share/classes/sun/security/ssl/ECDHKeyExchange.java:204]: Nit:  I think combining these lines would be < 80 chars

File: src/java.base/share/classes/sun/security/ssl/SSLSocketInputRecord.java
Comment (bradfordwetmore) [inline src/java.base/share/classes/sun/security/ssl/SSLSocketInputRecord.java:263]: Does the copyright need to get updated?

**Review by bradfordwetmore — COMMENTED** [review]
Didn't see any unit tests for the new methods.  Can approve after they are included.

File: src/java.base/share/classes/sun/security/ssl/SSLTrafficKeyDerivation.java
Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/SSLTrafficKeyDerivation.java:158]: Yes, I added 4 more whitespaces in line 158.

Comment (bradfordwetmore) [reply src/java.base/share/classes/sun/security/ssl/SSLTrafficKeyDerivation.java:158]: Ah, missed that.  Good to be consistent.

File: src/java.base/share/classes/sun/security/ssl/SSLSocketInputRecord.java
Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/SSLSocketInputRecord.java:263]: Oops, I missed this file.

**Comment by XueleiFan** [comment]
>

**Comment by XueleiFan** [comment]
> Update following for SSLPeerUnverifiedException -
> 
> https://github.com/openjdk/jdk/blob/master/src/java.naming/share/classes/com/sun/jndi/ldap/ext/StartTlsResponseImpl.java#L439

Hm, I will check more usage out of the JSSE implementation code.  Thank you!

**Comment by XueleiFan** [comment]
> Please consider adding simple tests to cover new APIs, similar to https://github.com/openjdk/jdk/pull/7705/files#diff-28e1041be519b6266b3b92aea29c5d8917c279880da7e5e9823a518196e96ea5

Thank you for the reference.  I will add some test cases.

File: src/java.base/share/classes/sun/security/ssl/ECDHKeyExchange.java
Comment (XueleiFan) [reply src/java.base/share/classes/sun/security/ssl/ECDHKeyExchange.java:204]: The exception name is too long to have in one line.  There are 83 characters in total if combining line 203 and 204.  3 characters exceeding 80 chars per line limit is not easy to tell with eyes.

Comment (RogerRiggs) [reply src/java.base/share/classes/sun/security/ssl/ECDHKeyExchange.java:204]: The 80 character recommendation isn't a hard limit, I favor readability in the 80-100 char range.

**Comment by XueleiFan** [comment]
ping ...

File: test/jdk/javax/net/ssl/ALPN/SSLServerSocketAlpnTest.java
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/ALPN/SSLServerSocketAlpnTest.java:487]: Copyright 2016->2022.

File: test/jdk/javax/net/ssl/ALPN/SSLSocketAlpnTest.java
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/ALPN/SSLSocketAlpnTest.java:483]: Copyright 2016->2022.

File: test/jdk/javax/net/ssl/SSLException/CheckSSLHandshakeException.java
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/SSLException/CheckSSLHandshakeException.java:2]: I am unsure about the copyrights on these files.  They are probably ok, but you should get approval by someone more familiar.

File: test/jdk/javax/net/ssl/SSLException/CheckSSLHandshakeException.java
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/SSLException/CheckSSLHandshakeException.java:32]: My personal preference would have been to combine all of these into a single testfile to minimize clutter in the logs and directories, but no strong objection.

File: test/jdk/javax/net/ssl/templates/SSLSocketSSLEngineTemplate.java
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/templates/SSLSocketSSLEngineTemplate.java:349]: Copyright update

File: test/jdk/javax/net/ssl/templates/SSLSocketTemplate.java
Comment (bradfordwetmore) [inline test/jdk/javax/net/ssl/templates/SSLSocketTemplate.java:544]: Copyright update

**Review by bradfordwetmore — APPROVED** [review]
Please see latest comments and let's verify the copyright is correct, but otherwise looks good.

File: test/jdk/javax/net/ssl/SSLException/CheckSSLHandshakeException.java
Comment (irisclark) [reply test/jdk/javax/net/ssl/SSLException/CheckSSLHandshakeException.java:2]: The location of the copyright is fine.  No need to enforce Oracle date formatting.

**Review by irisclark — APPROVED** [review]

**Comment by bradfordwetmore** [comment]
Thanks, @irisclark!

File: test/jdk/javax/net/ssl/SSLException/CheckSSLHandshakeException.java
Comment (XueleiFan) [reply test/jdk/javax/net/ssl/SSLException/CheckSSLHandshakeException.java:32]: I understand. I would like to have them in separated test cases so that it is easier to monitor the regressions in the future.

**Pull request closed without merging**
